lab-bundle-spec.md

```diff
@@ -158,41 +163,63 @@ environment_resources:
+attribute                        | required | type    | notes
+---------------------------------| -------- | ------- | --------------------------------------
+startup_script.type              |          | string  | The type of startup script. Only `deployment_manager` is supported.
+startup_script.path              |          | path    | Relative path to a directory tree with the script contents.
+startup_script.custom_properties |          | array   | Array of pairs. See below for details.
+cleanup_script.type              |          | string  | The type of cleanup script. Only `deployment_manager` is supported.
+cleanup_script.path              |          | path    | Relative path to a directory tree with the script contents.
+cleanup_script.custom_properties |          | array   | Array of pairs. See below for details.
+ssh_key_user                     |          | string  | If this project should use a user's SSH key, the id of that user.
 
 ```yml
 - type: gcp_project
   id: secondary_project
-  fleet: gcpfree
-  dm_template:
-    script: deployment_manager
+  variant: gcpfree
+  startup_script:
+    type: deployment_manager
+    path: dm_startup.zip
     custom_properties:
       - key: userNameWindows
         value: student
+  cleanup_script:
+    type: deployment_manager
+    path: dm_cleanup.zip
+    custom_properties:
+      - key: primary_project_zone
+        value: my_primary_project.zone
 ```
 
-> **NOTE:** Not all GCP fleet names are supported.
->
-> The existing concept of Qwiklabs' Fleets does not have a single analog in
-> content bundles. Notice that some fleet types map to resource types (e.g.
-> `gsuite_multi_tenant` fleet is now the `gsuite-domain` resource type).
-> However other fleets are "variants" of the same resource type (e.g.
-> "free", "ASL", and "standard" GCP projects). Therefore, not all fleet names that
-> an experienced Qwiklabs author may be familiar with, are allowed in the `fleet`
-> field.
+> **NOTE:** The existing concept of Qwiklabs' Fleets does not have a single
+> analog in content bundles.
 >
-> Presently, authors are allowed to specify one of the following fleet types for
-> a gcp_project:
-> - gcpd [default]
-> - gcpfree
-> - gcpasl
->
-> Future versions of the Content Bundle spec may use different terminology for
-> resource_type variations to avoid conflation.
+> Some fleet types map to resource types (e.g. `gsuite_multi_tenant` fleet is
+> now the `gsuite-domain` resource type), while other fleets are allowed as
+> "variants" of `gcp_project` (see below).
+
+###### Variants
+
+The allowed variants are:
+- gcpd [default]
+- gcpfree
+- gcpasl
+
+
+###### Custom properties
+
+attribute | required | type   | notes
+----------| -------- | ------ | --------------------------------------
+key       | ✓        | string | How the property will be referenced within the script.
+value     |          | string | A value to be passed into the script.
```

Comment:
Please clarify how this field is supposed to be used. Is this a fallback value? A constant to pass along at a later point?  Something else entirely?

Reply:
Good question. Now that I'm looking at it again, I'm not sure why this value couldn't just be hardcoded in the script. @dcbrandao ?

Reply:
Currently, authors use custom properties primarily for script re-usability. The same DM script can be used in multiple labs with variable tweaks.

You can check out the `custom_dm_properties` table in Gravitron to get a taste for how their used.

Reply:
But they still need to copy the file into their lab folder, right? Why wouldn't they just make the property replacement at that time?

Reply:
ping @jgan @dcbrandao

Reply:
It's a constant to be passed in, which allows a template to be reused across labs without having to laboriously replace all instances of one word with another when copying it around. It will also make it easier to enable true template reuse in the future.